polkadot/node/core/approval-voting/src/criteria.rs

```diff
@@ -170,14 +169,38 @@ fn relay_vrf_modulo_cores(
 		);
 	}
 
-	vrf_in_out
-		.make_bytes::<BigArray>(approval_types::v2::CORE_RANDOMNESS_CONTEXT)
-		.0
-		.chunks_exact(4)
-		.take(num_samples as usize)
-		.map(move |sample| CoreIndex(u32::from_le_bytes(clone_into_array(&sample)) % max_cores))
-		.unique()
-		.collect::<Vec<CoreIndex>>()
+	if 2 * num_samples > max_cores {
+		gum::error!(
+			target: LOG_TARGET,
+			n_cores = max_cores,
+			num_samples,
+			max_modulo_samples = MAX_MODULO_SAMPLES,
+			"Suboptimal configuration `num_samples` should be less than `n_cores` / 2",
+		);
+	}
+
+	let mut rand_chacha =
+		ChaCha20Rng::from_seed(vrf_in_out.make_bytes::<<ChaCha20Rng as SeedableRng>::Seed>(
+			approval_types::v2::CORE_RANDOMNESS_CONTEXT,
+		));
+
+	let mut res = HashSet::with_capacity(num_samples as usize);
+	while res.len() < min(num_samples, max_cores) as usize {
+		// Production networks should run with num_samples no larger than `n_cores/2`, so on worst
+		// case on each iteration this loop has >1/2 chance of finishing.
+		// With our current production parameters the chances for each iteration not hitting
+		// a duplicate are around 90%.
+		// See: https://github.com/paritytech/polkadot-sdk/pull/1178#discussion_r1376501206
+		// for more details.
+		for _ in 0..MAX_SAMPLING_ITERATIONS {
+			let random_core = rand_chacha.gen_range(0..max_cores).into();
```

Comment:
You mentioned past rand crate churn up thread, ala ChaCha=ChaCha20.  Afaik they've settled down, but..

If you worried there, then we can write this one out if you prefer.  I've just spent 20 minutes clicking through rustdoc, without actually finding their implementation.  lol
```
use rand_core::RngCore;
let r = rand_chacha.next_u32();
// Rejection sample for uniformity, but rejection is extremely unlikely.
let reject_threshold = (u32::MAX / max_cores) * max_cores;
if r > reject_threshold { next; }
let random_core = r % max_cores;
```

Reply:
Meh.  I'm being stupid there.  They cannot really change the meaning of a uniform sample.  lol